packages/engine-server/src/drivers/sqlite/tables/HierarchyTableUtils.ts

```diff
@@ -133,4 +133,33 @@ export class HierarchyTableUtils {
       }
     });
   }
+
+  /**
+   * Given a domain,
+   * recursively grab all descendants of the domain
+   * until we run out
+   *
+   * return all (parent, child) pairs
+   *
+   * more on recursive CTEs: https://www.sqlite.org/lang_with.html
+   */
+  public static getAllInDomain(
+    db: Database,
+    domainId: string
+  ): ResultAsync<HierarchyTableRow[], SqliteError> {
+    const sql = [
+      `WITH RECURSIVE "cte_hierarchy" (parent, child) AS (`, // 1. create a view (cte)
+      `  SELECT h.parent, h.child`, // 2. select a root note to anchor to. this is the domain note
+      `  FROM "Hierarchy" h`,
+      `  WHERE h.parent = "${domainId}"`, // 3. at this point cte_hierarchy has one row queued
+      `  UNION ALL`, // UNION ALL assuming we don't have dupes in the hierarchy table
+      `  SELECT h.parent, h.child`,
+      `  FROM "Hierarchy" h`, // 4. we are going to join the hierarchy table
+      `  JOIN "cte_hierarchy" c ON c.child = h.parent`, // 5. to the cte until we run out of children
```

Comment:
Can we add some infinite loop protection here? In case we have a loop between parent-child, child-parent somehow.

Reply:
Ideally we should do that check before inserting, but yeah. I'll add a check here for now